### PR TITLE
fix: validation change behavior regex match in `integer` and `numeric` to php function

### LINKED ISSUE #6489: Bug: "integer" validation rule 500 error
### PHP Version

8.1

### CodeIgniter4 Version

4.2.5

### CodeIgniter4 Installation Method

Manual (zip or tar.gz)

### Which operating systems have you tested for this bug?

Linux

### Which server did you use?

apache

### Database

_No response_

### What happened?

The validation rule "integer" fail as 500 error as below
```
{
    "title": "TypeError",
    "type": "TypeError",
    "code": 500,
    "message": "CodeIgniter\\Validation\\FormatRules::integer(): Argument #1 ($str) must be of type ?string, array given, called in /var/www/src/system/Validation/Validation.php on line 314",
    "file": "/var/www/src/system/Validation/FormatRules.php",
    "line": 132,
    "trace": [
        {
            "file": "/var/www/src/system/Validation/Validation.php",
            "line": 314,
            "function": "integer",
            "class": "CodeIgniter\\Validation\\FormatRules",
            "type": "->",
            "args": [
                [],
                null
            ]
        },
        {
            "file": "/var/www/src/system/Validation/Validation.php",
            "line": 163,
            "function": "processRules",
            "class": "CodeIgniter\\Validation\\Validation",
            "type": "->",
...
```

That validation allows string numeric (not integer) as well

### Steps to Reproduce

Validation rule used
```
protected $validationRules = [
        'integerAcceptsStringNumeric' => 'integer',
        'integerArrayError' => 'integer',
];
```

Data to validate
```
$data = [
        'integerAcceptsStringNumeric' => '1',
        'integerArrayError' => [],
];
```

### Expected Output

Validation fail and return 400 if the value is not an integer

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by ddevsr**
**Description**
Fixed #6489

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: system/Validation/FormatRules.php [hunk 1]
```diff
@@ -128,10 +128,12 @@ public function hex(?string $str = null): bool
...
     /**
      * Integer
+     *
+     * @param int|string|null $str
      */
-    public function integer(?string $str = null): bool
+    public function integer($str = null): bool
     {
-        return (bool) preg_match('/\A[\-+]?\d+\z/', $str ?? '');
+        return is_int($str);
```
Comment (iRedds) [inline system/Validation/FormatRules.php:136]: All data sent via HTTP is represented as strings and arrays. That is, an integer as a string will not pass validation.

**Review by iRedds — COMMENTED** [review]
Please add tests.

**Comment by iRedds** [comment]
I think that it would be more correct to remove typehint for passed values in all rules. Casting to the appropriate types within the method.

**Comment by kenjis** [comment]
This PR brings a lot of breaking changes. Probably unacceptable.

```
There was 1 error:

1) CodeIgniter\Validation\FormatRulesTest::testInteger with data set #0 (1, true)
TypeError: Argument 1 passed to CodeIgniter\Validation\FormatRulesTest::testInteger() must be of the type string or null, int given, called in /home/runner/work/CodeIgniter4/CodeIgniter4/vendor/phpunit/phpunit/src/Framework/TestCase.php on line 1545

/home/runner/work/CodeIgniter4/CodeIgniter4/tests/system/Validation/FormatRulesTest.php:867
phpvfscomposer:///home/runner/work/CodeIgniter4/CodeIgniter4/vendor/phpunit/phpunit/phpunit:97

--

There were 6 failures:

1) CodeIgniter\Validation\FormatRulesTest::testIntegerWithInvalidTypeData with data set "bool true" (true, true)
Failed asserting that false is identical to true.

/home/runner/work/CodeIgniter4/CodeIgniter4/tests/system/Validation/FormatRulesTest.php:812
phpvfscomposer:///home/runner/work/CodeIgniter4/CodeIgniter4/vendor/phpunit/phpunit/phpunit:97

2) CodeIgniter\Validation\FormatRulesTest::testInteger with data set #2 ('42', true)
Failed asserting that false is identical to true.

/home/runner/work/CodeIgniter4/CodeIgniter4/tests/system/Validation/FormatRulesTest.php:877
phpvfscomposer:///home/runner/work/CodeIgniter4/CodeIgniter4/vendor/phpunit/phpunit/phpunit:97

3) CodeIgniter\Validation\FormatRulesTest::testInteger with data set #3 ('-1', true)
Failed asserting that false is identical to true.

/home/runner/work/CodeIgniter4/CodeIgniter4/tests/system/Validation/FormatRulesTest.php:877
phpvfscomposer:///home/runner/work/CodeIgniter4/CodeIgniter4/vendor/phpunit/phpunit/phpunit:97

4) CodeIgniter\Validation\StrictRules\FormatRulesTest::testInteger with data set #0 ('0', true)
Failed asserting that false is identical to true.

/home/runner/work/CodeIgniter4/CodeIgniter4/tests/system/Validation/StrictRules/FormatRulesTest.php:876
phpvfscomposer:///home/runner/work/CodeIgniter4/CodeIgniter4/vendor/phpunit/phpunit/phpunit:97

5) CodeIgniter\Validation\StrictRules\FormatRulesTest::testInteger with data set #1 ('42', true)
Failed asserting that false is identical to true.

/home/runner/work/CodeIgniter4/CodeIgniter4/tests/system/Validation/StrictRules/FormatRulesTest.php:876
phpvfscomposer:///home/runner/work/CodeIgniter4/CodeIgniter4/vendor/phpunit/phpunit/phpunit:97

6) CodeIgniter\Validation\StrictRules\FormatRulesTest::testInteger with data set #2 ('-1', true)
Failed asserting that false is identical to true.

/home/runner/work/CodeIgniter4/CodeIgniter4/tests/system/Validation/StrictRules/FormatRulesTest.php:876
phpvfscomposer:///home/runner/work/CodeIgniter4/CodeIgniter4/vendor/phpunit/phpunit/phpunit:97
```

**Comment by kenjis** [comment]
> I think that it would be more correct to remove typehint for passed values in all rules. Casting to the appropriate types within the method.

Yes. When we introduce `declare(strict_types=1)`, many TypeError would occur.

**Comment by ddevsr** [comment]
> > I think that it would be more correct to remove typehint for passed values in all rules. Casting to the appropriate types within the method.
> 
> Yes. When we introduce `declare(strict_types=1)`, many TypeError would occur.
image

@kenjis this PR stuck on cache test php unit, i already adjusted in [`integerProvider`](https://github.com/codeigniter4/CodeIgniter4/pull/6490/commits/20eaa5378db1e29b7304172c2f906d1d111c91cc)

**Pull request closed without merging**